File: U1/TI1.4 RESTFULL_DOTNET_SINBDD_GR01/01 SERVIDOR/ConUni_Restfull_Dotnet_GR01/ConUni_Restfull_Dotnet_GR01/ec/edu/monster/ws/LongitudController.cs
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using ConUni_Restfull_Dotnet_GR01.ec.edu.monster.services;
using ConUni_Restfull_Dotnet_GR01.ec.edu.monster.models;
using ConUni_Restfull_Dotnet_GR01.ec.edu.monster.constants;
using ConUni_Restfull_Dotnet_GR01.ec.edu.monster.validators;

namespace ConUni_Restfull_Dotnet_GR01.ec.edu.monster.ws
{
    /// <summary>
 /// Controlador REST para conversiones de longitud
        /// Soporta conversiones entre: Milla, Metro, Pulgada
      /// </summary>
    [Route("api/[controller]")]
    [ApiController]
    [Produces("application/json")]
    public class LongitudController : ControllerBase
  {
        private readonly LongitudService _longitudService;

        public LongitudController(LongitudService longitudService)
        {
       _longitudService = longitudService;
        }

        /// <summary>
     /// Convierte valores entre diferentes unidades de longitud
        /// </summary>
   /// <param name="request">Solicitud de conversión con valor, unidad origen y unidad destino</param>
      /// <returns>Resultado de la conversión con valores exactos y redondeados</returns>
        /// <response code="200">Conversión exitosa o error de validación</response>
        /// <remarks>
  /// Unidades soportadas: Milla, Metro, Pulgada (case-insensitive)
      /// 
    /// Ejemplo de solicitud:
      /// 
   ///     POST /api/Longitud/convertir
     ///   {
       ///        "valor": "10",
  ///     "unidadOrigen": "Milla",
///        "unidadDestino": "Metro"
      /// }
      /// 
       /// </remarks>
   [HttpPost("convertir")]
   [ProducesResponseType(typeof(ConversionResultModel), StatusCodes.Status200OK)]
        [ProducesResponseType(typeof(ConversionResultModel), StatusCodes.Status400BadRequest)]
    public ActionResult<ConversionResultModel> Convertir([FromBody] ConversionRequest request)
        {
      // Normalizar las unidades a minúsculas para comparación
   var origen = request.UnidadOrigen.Trim().ToLower();
        var destino = request.UnidadDestino.Trim().ToLower();

        // Determinar qué conversión realizar basándose en origen y destino
   var resultado = (origen, destino) switch
     {
      ("milla", "metro") => _longitudService.ConvertirMillaAMetro(request.Valor),
    ("metro", "milla") => _longitudService.ConvertirMetroAMilla(request.Valor),
        ("milla", "pulgada") => _longitudService.ConvertirMillaAPulgada(request.Valor),
       ("pulgada", "milla") => _longitudService.ConvertirPulgadaAMilla(request.Valor),
      ("metro", "pulgada") => _longitudService.ConvertirMetroAPulgada(request.Valor),
    ("pulgada", "metro") => _longitudService.ConvertirPulgadaAMetro(request.Valor),

   // Conversión de una unidad a sí misma
   _ when origen == destino => HandleSameUnitConversion(request.Valor, request.UnidadOrigen, request.UnidadDestino),

     // Conversión no soportada
      _ => ConversionResultModel.Fallo(
      new ConversionErrorModel(
   ErrorConstants.ERROR_CONVERSION_LONGITUD,
    $"Conversión de {request.UnidadOrigen} a {request.UnidadDestino} no está soportada",
    ErrorConstants.TIPO_CONVERSION,
  TryParseDouble(request.Valor),
        request.UnidadOrigen,
         $"Las unidades soportadas son: {LongitudConstants.MILLA}, {LongitudConstants.METRO}, {LongitudConstants.PULGADA}"
    )
       )
            };

  return Ok(resultado);
        }

        /// <summary>
 /// Maneja conversiones donde origen y destino son la misma unidad
        /// </summary>
        private ConversionResultModel HandleSameUnitConversion(string valorString, string unidadOrigen, string unidadDestino)
  {
 // Primero validar que el valor sea válido
    var error = BaseValidator.ValidarStringPositivo(valorString, unidadOrigen, out double valor);
            if (error != null)
     return ConversionResultModel.Fallo(error);

     return ConversionResultModel.Exito(
   new UnidadConversionModel(
    valor,
   valor,
         CapitalizarUnidad(unidadOrigen),
CapitalizarUnidad(unidadDestino),
     "Longitud",
   1.0
   )
        );
        }

    /// <summary>
        /// Intenta convertir string a double para casos de error
     /// </summary>
      private double? TryParseDouble(string value)
      {
  if (double.TryParse(value, out double result))
       return result;
   return null;
        }

      /// <summary>
    /// Capitaliza el nombre de la unidad (Primera letra en mayúscula)
     /// </summary>
        private string CapitalizarUnidad(string unidad)
   {
   if (string.IsNullOrWhiteSpace(unidad))
       return string.Empty;

     unidad = unidad.Trim().ToLower();
  return char.ToUpper(unidad[0]) + unidad.Substring(1);
      }
 }
}

File: U1/TI1.3 RESTFULL_JAVA_SINBDD_GR01/01 SERVIDOR/ConUni_Restfull_Java_GR01/src/java/ec/edu/monster/ws/LongitudController.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/WebServices/GenericResource.java to edit this template
 */
package ec.edu.monster.ws;

import ec.edu.monster.constants.ConversionConstants.Longitud;
import ec.edu.monster.models.ConversionRequest;
import ec.edu.monster.models.ConversionResultModel;
import ec.edu.monster.services.LongitudService;
import jakarta.inject.Inject;
import jakarta.ws.rs.Consumes;
import jakarta.ws.rs.POST;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.Produces;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;

/**
 * REST Web Service para conversiones de longitud
 * Endpoints disponibles (todos POST con JSON):
 * - /api/longitud - Conversión general (basada en unidadOrigen y unidadDestino)
 *
 * Unidades soportadas: Milla, Metro, Pulgada
 *
 * Ejemplo de request:
 * {
 *   "valor": "100",
 *   "unidadOrigen": "Metro",
 *   "unidadDestino": "Milla"
 * }
 *
 * @author jeffe
 */
@Path("longitud")
public class LongitudController {

    @Inject
    private LongitudService longitudService;

    /**
     * Endpoint POST para convertir entre unidades de longitud
     * @param request ConversionRequest con valor, unidadOrigen y unidadDestino
     * @return Response con ConversionResultModel en JSON
     */
    @POST
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    public Response convertir(ConversionRequest request) {
        ConversionResultModel resultado;

        String origen = request.getUnidadOrigen();
        String destino = request.getUnidadDestino();
        String valor = request.getValor();

        // Determinar qué conversión realizar basándose en origen y destino
        if (Longitud.MILLA.equalsIgnoreCase(origen) && Longitud.METRO.equalsIgnoreCase(destino)) {
            resultado = longitudService.convertirMillaAMetro(valor);
        } else if (Longitud.METRO.equalsIgnoreCase(origen) && Longitud.MILLA.equalsIgnoreCase(destino)) {
            resultado = longitudService.convertirMetroAMilla(valor);
        } else if (Longitud.MILLA.equalsIgnoreCase(origen) && Longitud.PULGADA.equalsIgnoreCase(destino)) {
            resultado = longitudService.convertirMillaAPulgada(valor);
        } else if (Longitud.PULGADA.equalsIgnoreCase(origen) && Longitud.MILLA.equalsIgnoreCase(destino)) {
            resultado = longitudService.convertirPulgadaAMilla(valor);
        } else if (Longitud.METRO.equalsIgnoreCase(origen) && Longitud.PULGADA.equalsIgnoreCase(destino)) {
            resultado = longitudService.convertirMetroAPulgada(valor);
        } else if (Longitud.PULGADA.equalsIgnoreCase(origen) && Longitud.METRO.equalsIgnoreCase(destino)) {
            resultado = longitudService.convertirPulgadaAMetro(valor);
        } else {
            // Conversión no soportada
            return Response
                .status(Response.Status.BAD_REQUEST)
                .entity("{\"error\": \"Conversión no soportada. Unidades válidas: Milla, Metro, Pulgada\"}")
                .build();
        }

        // Retornar respuesta según el resultado
        if (resultado.isExitoso()) {
            return Response.ok(resultado).build();
        } else {
            return Response
                .status(Response.Status.BAD_REQUEST)
                .entity(resultado)
                .build();
        }
    }
}
